### PR TITLE
Move ismine to the wallet module

### PR DESCRIPTION
**Pull request by achow101**
`IsMine` isn't used outside of the wallet except for the tests. It also doesn't make sense to be outside of the wallet. This PR moves `IsMine` into the wallet module and for it to take a `CWallet` instead of `CKeyStore`. The test that used `IsMine` is also moved to the wallet tests.

This is first [prerequisites](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Class-Structure-Changes#ismine) for the wallet structure changes.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK. I think you'd need to update the linter after the rename to `CWallet`?

**Comment by achow101** [comment]
Added the circular dependency to the linter. It will be removed in the future once `ScriptPubKeyManager` is introduced.

**Comment by meshcollider** [comment]
Approach ACK, I'm ok with adding the circular dependency temporarily to be removed in a future PR as discussed with you at coredev

**Comment by practicalswift** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
ACK 3d77d8f48cc494fe5f272c175663b554e8aa6e3a (checked move-only and CWallet rename)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 3d77d8f48cc494fe5f272c175663b554e8aa6e3a (checked move-only and CWallet rename)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUikRQwAs9jDnCyIA93DlExBgTpbLV4DtTLKwfchRLURdMwt9/RFGjJAF9SDwwj1
gQC7eSmJYIrBff4Lx6BcYu/7vbqrSsxVmmHMQZJGZ8iv2hIjUg+YBwV6xgItRLbJ
hab8JXQ6a9QREK1beMlnvuWviIFX2EwEgk/dihByTIl7BLGL40x9t0XHuDWwdYz2
4VNsAYI6CO4oo0MztvpMKWPJutNeZpHJz6dBCj+ZB+Ra46yTZP6+HBLVHesZuomE
jjBWLfIHbv0/U8AkdhJCb+Y7uAKR9NNn6j2fGhuNqno/PhrFV4/oT9wa5XU7yLOP
Ze+0CYJa/j3oemkmWrYwWQRZyUcfdgK3h0nEMhWN36TpUalcbdR8+lynmxi5p82m
4YvlhiN9B2Hby4gXnVqnuNFqpi3HUfsexM0bxEJaoqg/Vn7A8lcnaf47e8WC5RYl
k9AeC/VV8qC6g6hYO+l4nLPs+nhkJtcdYFfwX+JxLs/wn/RJMFmvEMzdj6YloKx0
BLikxgFs
=j1at
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `8f465f280a7c3d1fd7c5da64fad29ace14155377e17cdb1af8661e6bfe1a967a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401088f465f280a7c3d1fd7c5da64fad29ace14155377e17cdb1af8661e6bfe1a967af010bfd3e02b8eb8e3044c59ab6f6ff6948b08fff0102c68689807eb4f6a02650ebff2ac48cd08f1045d080779f008895e65c3320adb3c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01077196549b6ca0901b97bc5ec3e9ce8b508f020ec25a59b7ad5a533f219af481883d4bef7936d7ad5e924b3bc695a27bb1d157108f1045d080779f0088b84c05d28e8daa10083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a22ec16de0ec4b2c0ee030c164e509e808f01035457cc7f5628599d7ef1991a457f19108f120cf4dd246cf20bafbbc645d606797fdd65326f83a05869f2b88efd533e91bca9208f1045d080779f008b6a43fd74f72194a0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010c71852d21f76846e9a13b79b068a396408f0205098a21fe7d8f770b28ec37ad8d01b25401532c088cacd87e1dc660c57de431308f1045d080779f008c70a97b983e1b5090083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/interfaces/wallet.h [hunk 1]
```diff
@@ -25,7 +24,10 @@ class CCoinControl;
 class CFeeRate;
 class CKey;
 class CWallet;
+enum isminetype : unsigned int;
 enum class FeeReason;
+typedef uint8_t isminefilter;
```
Comment (Empact) [inline src/interfaces/wallet.h]: nit: `using`?

Comment (MarcoFalke) [reply src/interfaces/wallet.h]: Yeah, but should change it in the original header file as well.

**Comment by Empact** [comment]
Concept ACK maybe keep `CKeystore` to avoid circular dependency?

**Comment by achow101** [comment]
> maybe keep `CKeystore` to avoid circular dependency?

In this whole wallet restructure, I don't think there's any way around a circular dependency here temporarily. IMO it makes more sense to do the change here rather than in another PR.

**Review by ryanofsky — COMMENTED** [review]
utACK 3d77d8f48cc494fe5f272c175663b554e8aa6e3a. Easy review with `--color-moved=dimmed_zebra`. 

It'd be nice to add a [project](https://github.com/bitcoin/bitcoin/projects/) to link the various issues and prs together (#16226, #16227, #15764, #16165)

**Review by ryanofsky — APPROVED** [review]

**Comment by MarcoFalke** [comment]
This will be merged next week, unless there are objections.

**Comment by promag** [comment]
ACK 3d77d8f48cc494fe5f272c175663b554e8aa6e3a.

**Review by sipa — APPROVED** [review]
Concept and superficial code review ACK 3d77d8f48cc494fe5f272c175663b554e8aa6e3a. I haven't verified move-only-ness.

Circular dependencies are expected while things are being moved around.

**Comment by jnewbery** [comment]
Concept ACK

**Comment by achow101** [comment]
Rebased

**Comment by MarcoFalke** [comment]
re-ACK e61de6306f (only change is rebase with git auto-merge)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK e61de6306f (only change is rebase with git auto-merge)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhq9QwAycC7mPLbHS+867UvtogkH0L4oG05HtDBKb0kCQ2e/CeLV2PaJB0s1cgo
hsZKIKtGErRyY1AnFq30VdkVweUD8UkICSb9q2zqPWYOXmG87hM4kOoySXlUCp/i
VaDPpd/nG1evVqkrfe3Q1WeQzgELJaf5zNghka1rcuJH/SMGVKTLY8zR/9KvG7n2
CQxyX6QbU6bddDqO3AfnE2+hDs5OpfPI3DXouz+pd17qLzH0wpksI+6CcnzI9w2J
x8jlQwZO8HB0FV0FB2N0rS3kqgfON9At3P++sGfdxxbfjZgyfxKtjKs54RHwSix5
UQ/P4c25iRqNlG3DI7+icZ1zp/JRNlZ5v20a/pIG/r/ZD6jBKn/MvY2aPZve9X8b
zfwvYMivFY7D8bHLWZ+3AqrDwZN/jDTuRhXAuXPzPxtWD6sTimAb95Yt/jqI67Pr
9ri/xaDHugu2Ay7wJM36MqSd9kV6cppXQXaeRKbaPHtO2QAAQApo/hlx4Kr6YJ1y
c0iF595K
=Cy/u
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `422247e430c86078dea26fdbc2e84832a7979645f33ca2b8bd2dd37469561631  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108422247e430c86078dea26fdbc2e84832a7979645f33ca2b8bd2dd37469561631f0103486ef9d7826a0665daf11314e2251b908fff0106d15c33eba2cd4be682e90350cfaf71808f1045d0b73c8f0088f8840bae7d4f9200083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01092dbb9113da58c3dfc809b0adfccf32e08f1045d0b73c8f008afd3553587fedd930083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010e06e17b5673128b64869011d56eb80f408f1045d0b73c8f008e598784ab54810130083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010f3d2509906d4901f7df5fd5db6328de408f1045d0b73c8f0081e81a32039c8fbff0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by meshcollider** [comment]
Very light code review ACK https://github.com/bitcoin/bitcoin/pull/16226/commits/e61de6306fd89fe9aae90253062e7b1b20343f8a

**Comment by laanwj** [comment]
posthumous ACK e61de6306fd89fe9aae90253062e7b1b20343f8a, agree it belongs with the wallet and that this is a better grouping than 'script', which is not a very useful axis to split (consensus/policy/wallet is better).

File: src/wallet/ismine.h [hunk 2]
```diff
@@ -28,8 +28,8 @@ enum isminetype
 /** used for bitflags of isminetype */
 typedef uint8_t isminefilter;
 
-isminetype IsMine(const CKeyStore& keystore, const CScript& scriptPubKey);
-isminetype IsMine(const CKeyStore& keystore, const CTxDestination& dest);
+isminetype IsMine(const CWallet& wallet, const CScript& scriptPubKey);
```
Comment (jnewbery) [inline src/wallet/ismine.h]: It's a shame that the declaration and definition have different argument names (`wallet` here in the decl and `keystore` in the def)

File: src/wallet/test/ismine_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    uncompressedKey.MakeNewKey(false);
+    CPubKey uncompressedPubkey = uncompressedKey.GetPubKey();
+    std::unique_ptr<interfaces::Chain> chain = interfaces::MakeChain();
+
+    CScript scriptPubKey;
+    isminetype result;
+
+    // P2PK compressed
+    {
+        CWallet keystore(chain.get(), WalletLocation(), WalletDatabase::CreateDummy());
```
Comment (jnewbery) [inline src/wallet/test/ismine_tests.cpp]: I'd prefer all these variables are named `wallet` instead of `keystore`, since that's what they are!

**Review by jnewbery — COMMENTED** [review]
post-merge ACK e61de6306fd89fe9aae90253062e7b1b20343f8a

I think the naming in variables/defns should be updated to be accurate. @achow101 - would it interfere with your further refactors if I opened a PR to do that?

**Pull request merged**